File: 03/day03.py
#!/usr/bin/env python3

try:
    from math import prod
except ImportError:
    from functools import reduce
    def prod(iter):
        return reduce(lambda x, y: x * y, iter)


def main():
    with open("input", "r") as f:
        lines = f.read().strip().split("\n")

    spec = [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)]

    print(solve(lines, 3, 1))
    print(prod([solve(lines, x, y) for x, y in spec]))


def solve(lines, r_offset, d_offset):
    r = 0
    d = 0
    trees = 0
    while d < len(lines):
        trees += 1 if lines[d][r] == "#" else 0
        r = (r + r_offset) % len(lines[d])
        d += d_offset
    return trees


if __name__ == "__main__":
    main()

File: 03/day03.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	lines, err := readLines("input")

	if err != nil {
		os.Exit(1)
	}

	fmt.Println(solve(lines, 3, 1))

	p2 := 1
	p2 *= solve(lines, 1, 1)
	p2 *= solve(lines, 3, 1)
	p2 *= solve(lines, 5, 1)
	p2 *= solve(lines, 7, 1)
	p2 *= solve(lines, 1, 2)

	fmt.Println(p2)
}

func solve(lines []string, roffset int, doffset int) int {
	r := 0
	d := 0
	trees := 0
	for d < len(lines) {
		if lines[d][r] == '#' {
			trees++
		}

		r = (r + roffset) % len(lines[d])
		d += doffset
	}
	return trees
}

func readLines(path string) ([]string, error) {
	lines := []string{}

	file, err := os.Open(path)

	if err != nil {
		return lines, err
	}

	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	return lines, scanner.Err()
}
